Nj initial

njohner opened the pull request:
Module Description
-------------------- 
This is a module to extract the contents of a zip file and add them as objects in Plone. It works with `my.teamraum` and can be adapted for `opengever`.
The core of the module is constituted by:
* A `ZipExtracter` class which handles extraction of elements from a zip file.
  * Safe against zip bombs: Checks announced size of files to extract (from zip file header) and stops extracting if that size is exceeded. We can also set a total maximum data volume allowed to be extracted.
  * Safe against extraction outside of destination folder.
  * Controlled RAM usage: Uses a buffer to write to a temporary file.
  * Handles things like zip file containing several files with the same name.
* A `ZipExtractView` class used to show a preview of the extracted file structure and allowing selection of files to extract (showing file sizes)

The other elements are:
* An `IZip` interface to mark zip files.
  * Files are marked as zip on creation or modification
  * Here events are only available to mark `ftw.file` but can be easily extended to other content types (for opengever see the `nj_opengever` branch).
* An `IFolderCreator`and `IFileCreator` interface used to define adapters for file and folder creation in different contexts.
  * Implementations for `my.teamraum` are found in https://github.com/4teamwork/ftw.file/pull/73 and https://github.com/4teamwork/ftw.workspace/pull/89
* An `IFile` interface to standardise a few methods in file handling.
  * An implementation of the interface for `ftw.file` is also provided

Finally a file action was added to extract elements from a zip file in https://github.com/4teamwork/bumblebee.file/pull/38

resolves https://github.com/4teamwork/my.teamraum/issues/667

Commit:
Initial commit for a zip extractor module

njohner commented:
Open questions and still to do
------------------------------
* `bumblebee.file` now depends on `ftw.zipextract`. Would there be a better way for doing this?
* I implemented the events and `IFile` interface for `ftw.file` directly in `ftw.zipexrtact`. Is that OK or would you rather have them in `ftw.file`?
* I have not added any tests yet, I first wanted some feedback on the module itself.
* I've started adapting the module for `opengever`. There is an `nj_opengever` branch here and an `nj_zipextract` branch in `opengever.core`

jone commented on ftw/zipextract/File.py:
```diff
@@ -0,0 +1,27 @@
+from interfaces import IFile
+
+
+class _FileBase():
```
I would not prefix classes with an underscore. You could add a docstring indicating that this class is meant as abstract base class instead.
I'd probably also declare not implemented methods like this:
```python
def get_blob(self):
    raise NotImplementedError()
```

jone commented on ftw/zipextract/File.py:
```diff
@@ -0,0 +1,27 @@
+#from opengever.document.document import IDocumentSchema
+from zope.component import adapts
+from zope.interface import implements
+from interfaces import IFile
```
Please use absolut imports.

jone commented on ftw/zipextract/browser/templates/zipextract.pt:
```diff
@@ -0,0 +1,22 @@
+        <form tal:attributes="action string:${here/absolute_url}/zipextract">
+            <div tal:content="structure view/tree_html" ></div>
+            <div>
+            <input type="checkbox" id="create_root_folder_checkbox" name="create root folder" checked="checked" /><label for="create_root_folder_checkbox">Create root folder</label></div>
```
The label is not translated.

jone commented on ftw/zipextract/browser/templates/zipextract.pt:
```diff
@@ -0,0 +1,22 @@
+    Extract files
+    </metal:block>
+    <metal:block fill-slot="main">
+        <form tal:attributes="action string:${here/absolute_url}/zipextract">
```
I'd use `method="POST"`. You don't want the values to end up in the URL (as it happens with `GET`)

jone commented on ftw/zipextract/browser/zipextract_view.py:
```diff
@@ -0,0 +1,71 @@
+
+    def __call__(self):
+        self.zipextracter = ZipExtracter(self.context)
+        self.build_tree_html()
```
This does not need to be done when the form is submitted, right?

njohner replied:
No indeed... I wanted to ask about this but forgot. How do I do this properly? when the form is submitted, it seems that a new `ZipExtractView` is created and then the `__call__` method gets called...

njohner replied:
Ok nevermind, I misunderstood. `self.zipextracter` gets generated anew when the form is submitted, but that cannot be avoided. The html tree is now handled by the template so that is not an issue anymore.

jone commented on ftw/zipextract/browser/zipextract_view.py:
```diff
@@ -0,0 +1,71 @@
+            node = node.subtree[k]
+        return node.file_dict[keys[-1]]
+
+    def redirect_to_workspace(self):
```
Id use "container" instead of "workspace" as it may be something else (e.g. a dossier).

njohner replied:
done

jone commented on ftw/zipextract/browser/zipextract_view.py:
```diff
@@ -0,0 +1,71 @@
+        for file_node in folder_node.file_dict.values():
+            label = file_node.name + ": " + self._file_size_repr(file_node)
+            self.tree_html += '<li id="{id}" class="file"><input type="checkbox" value="{path}"\
+                id="{id}_checkbox" name="files:list" /><label for="{id}_checkbox">{name}</label></li>\n'\
```
I'd try to move rendering to a template. Doing it like this is prone to prone for XSS-bugs.

njohner replied:
done

jone commented on ftw/zipextract/browser/zipextract_view.py:
```diff
@@ -0,0 +1,71 @@
+        if not request.get('extract'):
+            return
+        if request.get('extract') == "Extract all":
+            file_list=None
```
Please run `bin/pycodestyle` 😉

jone commented on ftw/zipextract/configure.zcml:
```diff
@@ -0,0 +1,50 @@
+    <configure zcml:condition="installed ftw.file">
+        <adapter
+            for="ftw.file.content.file.IFile"
+            factory=".File.FtwFile" />
```
I'd use lowercase filenames (`file.py` instead of `File.py`).

njohner replied:
done

jone commented on ftw/zipextract/configure.zcml:
```diff
@@ -0,0 +1,50 @@
+
+    <configure zcml:condition="installed ftw.file">
+        <adapter
+            for="ftw.file.content.file.IFile"
```
I'd try to implement a "archetypes" default implementation. I think it should be possibl to not make it `ftw.file` specific while still making it work with `ftw.file`.
We can rely on the file-field beeing defined as primary field, so that we can do thinks like `context.getPrimaryField().getBlob()` and such.

If it needs to be `ftw.file`-specific, it belongs in the `ftw.file`-package. In this case the `ftw.file`-package declares a `tests`-dependency to `ftw.zipextract`.

njohner replied:
Done. Made a default implementation for archetypes and for dexterity. There is no need to do anything in `ftw.file`

jone commented on ftw/zipextract/events.py:
```diff
@@ -0,0 +1,8 @@
+from interfaces import IFile
+from interfaces import IZip
+from Products.Five.utilities.marker import mark
```
For letting an object provide an interface we should use `zope.interface.alsoProvides` instead of `mark`.

njohner replied:
Got rid of the interface altogether

jone commented on ftw/zipextract/events.py:
```diff
@@ -0,0 +1,8 @@
+
+def set_zip_on_creation(file, event):
+    if IFile(file)._is_zip():
+        mark(file, IZip)
```
I'm not yet happy with the approach of marking ZIPs.
One disadvantage is that we then need to process all files in an upgrade and apply the interface.

I'd prefer to not have this interface.
If it is only for the bumblebee action, I'd test the mimetype directly in the view.

njohner replied:
Got rid of the marker interface. I'm simply testing the mimetype when generating the bumblebee actions. I'm also testing the mimetype before extracting from a zip (in case someone calls the action directly via URL).

jone commented on ftw/zipextract/interfaces.py:
```diff
@@ -0,0 +1,28 @@
+    pass
+
+
+class ObjectCreatorBase():
```
Since this is not an interface (it does not subclass `Interface`), I'd move that to another file.

njohner replied:
done

jone commented on ftw/zipextract/profiles/default/metadata.xml:
```diff
@@ -0,0 +1,6 @@
+<?xml version="1.0"?>
+<metadata>
+    <dependencies>
+        <dependency>profile-ftw.upgrade:default</dependency>
```
`ftw.zipexport` should not install `ftw.upgrade`.
For declaring upgrade steps it is not necessary to install the generic setup profile of `ftw.upgrade`; so it is the job for policy packages / projects to install it.

jone reviewed (commented):
I'd could yet review completely, but have some inputs already.

Also:
- It would be great to have a good readme at the end, also describing things that need to be implemented (e.g. when someone does not use our packages) and maybe screenshots.
- Screenshots in the pull request would help visualize the solution as a reviewer
- We could also think about also supporting Plone 5 from the beginning.

jone commented:
Regarding Questions:

> bumblebee.file now depends on ftw.zipextract. Would there be a better way for doing this?

`bumblebee.file` should not depend on `ftw.zipexport`, but we can declare an `extras` in the `setup.py` of `bumblebee.file` and build it robust so that it enables the feature when it detects that `ftw.zipextract` is installed too.

> I implemented the events and IFile interface for ftw.file directly in ftw.zipexrtact. Is that OK or would you rather have them in ftw.file?

I already made comments about that:
- IZip: I hope we can get rid of that, so that we have no event handlers at all.
- Extraction: I think we can do that with a default Archetypes implementation, so that we do not need a ftw.file-specific implementation.
- Creation: we still need to tell the portal_type of folders and files. I think an adapter is good for that; both are already in `ftw.workspace`. I think we can keep that.

> I have not added any tests yet, I first wanted some feedback on the module itself.

I'd add tests for all components so that we can detect bugs introduced by refactorings.

Commit:
Add default implementations for Archetypes.

Commit:
Renamed redirect_to_workspace to redirect_to_container.

Commit:
Add error handling to zipextracter.

Commit:
Add tests for zipextracter.

Commit:
Remove IZip interface and associated events.

Commit:
Add default implementation and tests for Dexterity types.

njohner commented:
I also added tests in `ftw.zipexrtact` and in `bumblebee.file`.

Commit:
Split archetype from dexterity implementation.

Commit:
Add translations.

njohner commented:
After refactoring we have:

Module Description
-------------------- 
This is a module to extract the contents of a zip file and add them as objects in Plone. It works in a standard plone site and in `my.teamraum`. It could be adapted for `opengever`.
The core of the module is constituted by:
* A `ZipExtracter` class which handles extraction of elements from a zip file.
  * Safe against zip bombs: Checks announced size of files to extract (from zip file header) and stops extracting if that size is exceeded. We can also set a total maximum data volume allowed to be extracted.
  * Safe against extraction outside of destination folder.
  * Controlled RAM usage: Uses a buffer to write to a temporary file.
  * Handles things like zip file containing several files with the same name.
* A `ZipExtractView` class used to show a preview of the extracted file structure and allowing selection of files to extract (showing file sizes)

The other elements are:
* An `IFolderCreator`and `IFileCreator` interface used to define adapters for file and folder creation in different contexts. The module contains default implementations for Archetype and Dexterity types.
  * For `my.teamraum` we need adapters for the folder creation, which are found in https://github.com/4teamwork/ftw.workspace/pull/89
* An `IFile` interface to standardise a few methods in file handling.
  * We provide an implementation of the interface for Archetypes and Dexterity types.

To have a button for the extraction in `my.teamraum`, a file action was added in `bumblebee.file`: https://github.com/4teamwork/bumblebee.file/pull/38

resolves https://github.com/4teamwork/my.teamraum/issues/667

Commit:
Remove file extension from root directory name.

Commit:
Added module description to README.

njohner commented:
The interface looks like this in `my.teamraum`:
-----------------------------------------------
**For files that are not zip files, there is no *extract* button**
![no_extract_button](https://user-images.githubusercontent.com/7374243/34841386-685d8774-f708-11e7-8675-ef44e253dddc.png)
**The *extract files* button is available for zip files in the standard view**
![screen shot 2018-01-15 at 11 03 10](https://user-images.githubusercontent.com/7374243/34937063-c2719060-f9e3-11e7-82fb-d140b29301df.png)
**Clicking the *extract files* button takes you to the `zipextract_view`. Here you get a preview of the contents of the zip file and can choose which files to extract and whether to generate a root folder (yes by default).**
![screen shot 2018-01-15 at 11 04 14](https://user-images.githubusercontent.com/7374243/34937153-11483ff4-f9e4-11e7-9546-2e44a87c56c4.png)


**After extraction we get the root folder named after the file**
![extracted_contents_root](https://user-images.githubusercontent.com/7374243/34841411-81b74368-f708-11e7-87bf-cfadaa18207d.png)

**In the root folder you find the file structure that you extracted**
![extracted_contents](https://user-images.githubusercontent.com/7374243/34841416-84c1f972-f708-11e7-9349-c693909ca5fa.png)

Commit:
Move tree representation to template.

Commit:
Add header to the zip extraction view.

Commit:
Adapt tests for tree depth of 3.

Commit:
Move warnings to log infos.

Commit:
Code style improvements.

njohner commented:
@jone @maethu 
Ok this is ready for re-review.

jone commented on README.rst:
```diff
@@ -0,0 +1,86 @@
+==============
+
+* The ``ftw.zipextract.zipextracter.ZipExtracter`` class handles the extraction from the zip file.
+* The ``ftw.zipextract.browser.zipextract_view.ZipExtractView``is used for the rendering of the extraction view. It is registered as ``zipextract`` and will be applied on its context.
```
Missing space after ```ZipExtractView`` ```

jone commented on ftw/zipextract/browser/templates/macros.pt:
```diff
@@ -0,0 +1,29 @@
+
+    <metal:macro-block metal:define-macro="show_tree">
+        <tal:repeat-file tal:repeat="file_node folder_node/file_dict/values">
+            <li class="file"  tal:define="fid python:view.norm_id(file_node); label python:view.file_repr(file_node); path file_node/path; chid python:fid+'_checkbox' " tal:attributes="id fid">
```
Breaking the define into multiple lines would make it better readable (especially on github :wink:):
```html
            <li class="file" tal:define="fid python:view.norm_id(file_node);
                                         label python:view.file_repr(file_node);
                                         path file_node/path;
                                         chid python:fid+'_checkbox' "
                             tal:attributes="id fid">
```

jone commented on ftw/zipextract/browser/templates/macros.pt:
```diff
@@ -0,0 +1,29 @@
+            <li class="file"  tal:define="fid python:view.norm_id(file_node); label python:view.file_repr(file_node); path file_node/path; chid python:fid+'_checkbox' " tal:attributes="id fid">
+                <input  type="checkbox" name="files:list" tal:attributes="value path; id chid"/>
+                <label tal:attributes="for chid ; value label">
+                    <span tal:replace=label></span>
```
Since this is XML, I suggest to form a habit to use quotes 😉
```html
 <span tal:replace="label" />
```
Since you are actually filling the `label`-tag, you could also use `tal:content` on the `label`-tag:
```html
<label tal:attributes="for chid;
                       value label"
       tal:content="label" />
```
The result is the same as replacing the `span`-tag.

jone commented on ftw/zipextract/browser/templates/zipextract.pt:
```diff
@@ -0,0 +1,33 @@
+        <form tal:attributes="action string:${here/absolute_url}/zipextract" method="POST">
+
+            <ul class="zipextract file_tree" tal:define="folder_node view/zipextracter/file_tree" >
+                <metal:block tal:define="folder_node view/zipextracter/file_tree" metal:use-macro='context/@@zipextract.macros/show_tree'>
```
For consistency, I'd always use double-quotes:
```html
<metal:block tal:define="folder_node view/zipextracter/file_tree"
             metal:use-macro="context/@@zipextract.macros/show_tree" />
```

jone commented on ftw/zipextract/browser/zipextract_view.py:
```diff
@@ -0,0 +1,63 @@
+from Products.Five.browser import BrowserView
```
Please run `bin/pyflakes`; it will show you that this import is not used at the moment, and probably more 😉

jone commented on ftw/zipextract/browser/zipextract_view.py:
```diff
@@ -0,0 +1,63 @@
+from z3c.form import form
+
+
+class ZipExtractView(form.Form):
```
What is the advantage of using `z3c.form` in this situation over simply using a `BrowserView`?
The z3c.form usually helps when defining a schema with `zope.schema` and then let it automatically be rendered.
Since you have no schema, no validation and build the HTML with your own template I don't see the benefit and would therefore go with `BrowserView` and `__call__` 😉

njohner replied:
None indeed. I got confused at some point and misinterpreted your comment https://github.com/4teamwork/ftw.zipextract/pull/1#discussion_r160420777 and thought we wanted to avoid having to rebuild the `zipextracter` object in when doing the extraction. I tried to use buttons for that, to avoid having to use the `__call__` method when doing the extraction. I realize that was really foolish 👎. Anyway, bottom line is, I'm back to using the BrowserView (hence the unused import too...)

jone commented on ftw/zipextract/browser/zipextract_view.py:
```diff
@@ -0,0 +1,63 @@
+        return file_node.name + ": " + self._file_size_repr(file_node)
+
+    @staticmethod
+    def norm_id(node):
```
This function is called from the template.
It is usually a better design to pass the data already prepared into the template, when possible.
You wouldn't have to call things of the view at alle, then.

njohner replied:
I'd still have to get the data from the view though, so what difference does this make? Is it just cosmetics?

jone replied:
This is a static method, so we have no dependency to the view (`self`).

I'd move this method to the `_FileNode` class; it seems to belong there, not in the view.
In the template you can then replace `python:view.norm_id(file_node)` with `file_node/id` or such; isn't that better?

njohner replied:
Yes, that's actually what I've done in the meantime, just not committed yet. I was just wondering if there was a technical reason why calling from the template was to be avoided. But of course conventions, readability and simply trying to write nice code is reason enough.
When I commit you'll see I now also avoid calling the `file_repr` method from within the template.

jone commented on ftw/zipextract/file.py:
```diff
@@ -0,0 +1,74 @@
+from zope.interface import implements
+
+
+class FileBase():
```
In python 2.x, you are creating an old-style-class here. This has impacts such as that you cannot use `super`.
We always want to have new-style-classes.

Example:
```python
Python 2.7.10 (default, Oct 23 2015, 19:19:21)
[GCC 4.2.1 Compatible Apple LLVM 7.0.0 (clang-700.0.59.5)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> class OldStyle(): pass
...
>>> class NewStyle(object): pass
...
>>> type(OldStyle)
<type 'classobj'>
>>> type(NewStyle)
<type 'type'>
>>> dir(OldStyle)
['__doc__', '__module__']
>>> dir(NewStyle)
['__class__', '__delattr__', '__dict__', '__doc__', '__format__', '__getattribute__', '__hash__', '__init__', '__module__', '__new__', '__reduce__', '__reduce_ex__', '__repr__', '__setattr__', '__sizeof__', '__str__', '__subclasshook__', '__weakref__']
>>>
```
See also: https://stackoverflow.com/questions/4015417/python-class-inherits-object

njohner replied:
Oh thanks! I saw a few instances in `class(object)` in the plone code recently and had no idea why this was done...

jone commented on ftw/zipextract/implementations_at.py:
```diff
@@ -0,0 +1,19 @@
+
+class ATFileCreator(ObjectCreatorBase):
+    implements(IFileCreator)
+    adapts(IATFile)
```
How is this adapter adapting an IATFile? It is a factory adapter, so it will create the file. At the point the adapter is created it cannot yet exist.

njohner replied:
Yes I know this is weird and I'd be happy about a better solution. For the `FolderCreator`s, the adapter takes the parent folder to know which type of folder we are working with, and hence what `portal_type` to create. For the `FileCreator`s, I use the current zip file (i.e. the context of the zipextracter) and adapt that. This seemed strange to me when I wrote it, but I couldn't think of another way.

jone replied:
Ah, now I understand it.

I propose to change it so that the adapted context is always the future parent. This would work with our use cases. The current implementation also does not support adding DX-objects within AT-objects or vice-versa.

For example in workspaces:
Files and folders can both be added within `Workspace`-objects or within `TabbedViewFolder`-objects. Therefore I'd let both adapt `Products.CMFCore.interfaces.IFolderish`. `IFolderish` is provided by all folderish Plone objects (AT and DX).

The problem with this change is that we then cannot distinguish AT and DX creator adapters, because we are adapting the parent. We also cannot adapt the future object since it is not yet created at this point.
I would solve this problem by also adapting the FTI (Factory Type Information) for the portal-type to be created (get it with `getToolByName(context, 'portal_types').get('File')`.

That means that we need to move the decision, which portal type to create, outside of this creator. We also may need to support creating multiple file types depending on the content. For example in standard Plone we could create `Image`-objects for images. When `ftw.mail` is installed we could create a mail-object for `.eml`-files.

So the adapters could look like that:
```python
@implementer(IFolderCreator)
@adapter(Interface, Interface, IDexterityFTI)
class DXFolderCreator(object):

    def __init__(self, container, request, fti):
        self.container = container
        self.request = request
        self.fti = fti

    def create(self, title):
        return createContentInContainer(self.container, self.fti.factory, title=title)


@implementer(IFileCreator)
@adapter(Interface, Interface, IDexterityFTI)
class DXFileCreator(object):

    def __init__(self, container, request, fti):
        self.container = container
        self.request = request
        self.fti = fti

    def create(self, filename, blobfile):
        obj = createContentInContainer(self.container, self.fti.factory, title=filename)
        field = IPrimaryFieldInfo(obj).field
        field.set(obj, NamedBlobFile(obj))
        return obj

    
@implementer(IFolderCreator)
@adapter(Interface, Interface, Interface)
class ATFolderCreator(object):

    def __init__(self, container, request, fti):
        self.container = container
        self.request = request
        self.fti = fti

    def create(self, title):
        normalized_id = normalize_id(name)
        chooser = INameChooser(context)
        new_id = chooser.chooseName(normalized_id, context)
        return self.container.invokeFactory(self.fti.factory, new_id, title=title)

    
@implementer(IFileCreator)
@adapter(Interface, Interface, Interface)
class ATFileCreator(object):

    def __init__(self, container, request, fti):

        self.request = request
        self.fti = fti

    def create(self, filename, blobfile):
        normalized_id = normalize_id(filename)
        chooser = INameChooser(context)
        new_id = chooser.chooseName(normalized_id, context)
        obj = self.container.invokeFactory(self.fti.factory, new_id, title=title)
        obj.getPrimaryField().set(obj, blobfile)
        return obj
```

Thoughts and hints:
- I combined creating and setting the blob file, because I find this better understandable.
- The DX-FTI provides IDexterityFTI, but the AT-FTI does not provide interfaces which are not provided by DX, that's why the AT adapters adapt "Interface" at this position.
- I also mixed the request in: this allows for easier customization with browser layers in the future.
- I'm using the primary field instead of hard-coded "file" field name; the primary field is usually the correct one.

When using the adapter you could first query for a named adapter, then fallback to the default
adapter. This will also allow us to e.g. customizate the "Mail" creator but not the "File" creator.

```python
def get_creator(interface, container, portal_type):
    fti = getToolByName(container, 'portal_types').get('portal_type')
    creator = queryMultiAdapter((container, container.REQUEST, fti), interface, name=portal_type)
    if creator is not None:
        return creator
    else:
        return getMultiAdapter((container, container.REQUEST, fti), interface)


get_creator(IFolderCreator, context, 'TabbedViewFolder')
get_creator(IFileCreator, context, 'File')
```

njohner replied:
Hum, I think we'll have to discuss that tomorrow. I modified the interfaces, but it does not really change the issue knowing what portal type will be created. Or maybe there is something I don't understand.
`IFolderCreator` and `IFileCreator` get invoked from `zipextracter`. And there I still only have the container where the new item will be created and a filename or foldername. So I still end up using the container's portal_type as argument of `get_creator`.

Also I was not able to use the `fti` for archetypes. `invokeFactory` expects a `portal_type` as first argument and `fti.portal_type` is always `Plone Site` no matter what the container was. And I don't know what to do with `ftw.factory` in that case. The way you wrote it `obj = self.container.invokeFactory(self.fti.factory, new_id, title=title)` will not work (as said it expects a `portal_type`).

PS: I know the last commit is still dirty with commented code and such.

jone commented on ftw/zipextract/tests/test_zipextract.py:
```diff
@@ -0,0 +1,172 @@
+
+class ZipExtracterTestBase(FunctionalTestCase):
+
+    def AddMultiZipFile(self):
```
Please use the underscore-style for method names.

jone commented on ftw/zipextract/zipextracter.py:
```diff
@@ -0,0 +1,289 @@
+    from zope.container.interfaces import INameChooser
+
+
+class _FileNode():
```
Please do not start classnames with underscores 😉

jone commented on ftw/zipextract/zipextracter.py:
```diff
@@ -0,0 +1,289 @@
+class _FileNode():
+
+    def __init__(self, parent_folder, filename, fileid, info):
+        self.name = _(filename)
```
Why are you trying to translate the filename?

njohner replied:
Hum, I think I wanted to escape the html for safety here and thought the `MessageFactory` also did that. Should I use something like `escape` from `xml.sax.saxutils`, because folder and file names are used in the template? Or is that  unnecessary and plone takes care of it?

jone replied:
The page tempalte engine takes care of that:
given `foo="a<b"`
safe: `<div tal:content="foo" />` => `<div>a&lt;b</div>`
not safe: `<div tal:content="structure foo" />` => `<div>a<b</div>`

So you want to use `structure` whenever you are 100% sure that the inserted value is valid and safe HTML. Normally you just do not add `structure` and you are safe when using `tal`-commands (content, replace, attributes, etc).

njohner replied:
Right, I'd seen that at some point... Thanks!

jone commented on ftw/zipextract/zipextracter.py:
```diff
@@ -0,0 +1,289 @@
+            err_msg = "{} is not a zip file, can't extract from it.".format(
+                context.absolute_url_path())
+            raise TypeError(err_msg)
+        file = ifile.get_blob()
```
`file` is a builtin name and should therefore never be used as a variable name.
You could use `file_io` or `blob_file` .

njohner replied:
🥇

jone commented on ftw/zipextract/zipextracter.py:
```diff
@@ -0,0 +1,289 @@
+        return file_list
+
+
+class ZipExtracter():
```
subclass `object`, as above

jone commented on ftw/zipextract/zipextracter.py:
```diff
@@ -0,0 +1,289 @@
+        self.file_creator = IFileCreator(self.context)
+
+    def _isdir(self, type):
+        if type == "Folder":
```
`type` is a builtin function and should not be redeclared.

njohner replied:
Not using this method anymore, I'll delete it

jone commented on ftw/zipextract/zipextracter.py:
```diff
@@ -0,0 +1,289 @@
+            if is_folder:
+                continue
+            keys = target_path.split(os.path.sep)
+            curr_node = self.file_tree
```
This method is quite long. I'd try to refactor it by extracting functionality into other methods.
I'd also try to only have one `for`-loop over the file infos.

njohner replied:
The two `for` loops were there because you can have files with the same name as folders inside zip files. But with how the tree is implemented now, this is not needed anymore. This shortens the method a lot, let me know if you still think I should try to delegate some functionality to other methods.

Commit:
Refactor IFileCreator and IFolderCreator adapters.

Commit:
Handle name conflicts in zip files.

jone committed:
Draft factory type decider.

Commit:
Finalize factory type decider

Commit:
Cleanup IFile interface

Commit:
Add tests for default implementations of interfaces

Commit:
Code cleanup

Commit:
Add tests for event handling.

Commit:
Update README file

Commit:
Remove dependency to ftw.upgrade

njohner commented:
@jone @maethu 
* I finished implementing the `FactoryTypeDecider`. Instead of returning a `portal_type`, I made it directly return a Factory type information instead. 
* I also added the mimetype as an argument to the `FileCreator.create` method. I think this makes sense if we have the mimetype as argument of the `FactoryTypeDecider`, and it might be useful for certain file types
* Tests to check that the default implementations match the interfaces
* Tests to check that extraction generates the proper events

So I think this is ready for a 3rd review.

jone committed:
Update readme (markup, typo) and setup.py.

jone committed:
Update translations.

jone committed:
Do not try to extract folder as file.

When selecting a folder (in order to select all files), it was
submitted as file too, which lead to an error in the extractor when it
tried to access the folder in the file list.

jone committed:
Respect typesUseViewActionInListings when redirecting.

jone committed:
Subclass from object.

jone reviewed (approved):
I have added some changes to this branch (typos, translations, small fixes).

Although the package is not complete I think we now have a solid ground to further build ontop and I would like to merge it as it is. 👍 

I started making issues for things I'd like to extend or change.

Commit:
Update french translations

njohner commented:
Perfect thanks. I just committed a small change to match your change in the German translation.
